dnf5-plugins/system_upgrade_plugin/system_upgrade.cpp

```diff
@@ -0,0 +1,517 @@
+    // Upgrade operation already removes all element that must be removed.
+    // Additional removal could trigger unwanted changes in transaction.
+    ctx.base.get_config().get_clean_requirements_on_remove_option().set(
+        libdnf5::Option::Priority::PLUGINDEFAULT, false);
+    ctx.base.get_config().get_install_weak_deps_option().set(libdnf5::Option::Priority::PLUGINDEFAULT, false);
+}
+
+void SystemUpgradeUpgradeCommand::run() {
+    auto & ctx = get_context();
+
+    if (!std::filesystem::is_symlink(get_magic_symlink())) {
+        throw libdnf5::cli::CommandExitError(0, M_("Trigger file does not exist. Exiting."));
+    }
+
+    const auto & symlinked_path = std::filesystem::read_symlink(get_magic_symlink());
+    if (symlinked_path != get_datadir()) {
+        throw libdnf5::cli::CommandExitError(0, M_("Another upgrade tool is running. Exiting."));
+    }
+
+    std::filesystem::remove(get_magic_symlink());
+
+    auto state = get_state();
+    if (state.get_data().status != STATUS_READY) {
+        throw libdnf5::cli::CommandExitError(1, M_("Use `dnf5 system-upgrade reboot` to begin the upgrade."));
+    }
+
+    const auto & transaction_json_path = std::filesystem::path(get_datadir()) / "system-upgrade-transaction.json";
+    auto transaction_json_file = libdnf5::utils::fs::File(transaction_json_path, "r");
+
+    const auto & json = transaction_json_file.read();
+    transaction_json_file.close();
+
+    const auto & goal = std::make_unique<libdnf5::Goal>(ctx.base);
+
+    auto settings = libdnf5::GoalJobSettings();
+    goal->add_serialized_transaction(json, settings);
+
+    auto transaction = goal->resolve();
+    if (transaction.get_problems() != libdnf5::GoalProblem::NO_PROBLEM) {
+        throw libdnf5::cli::GoalResolveError(transaction);
+    }
+
+    auto callbacks = std::make_unique<PlymouthTransCB>();
+    /* callbacks->get_multi_progress_bar()->set_total_num_of_bars(num_of_actions); */
+    transaction.set_callbacks(std::move(callbacks));
+
+    const auto result = transaction.run();
+    std::cout << std::endl;
+    if (result != libdnf5::base::Transaction::TransactionRunResult::SUCCESS) {
+        std::cerr << "Transaction failed: " << libdnf5::base::Transaction::transaction_result_to_string(result)
+                  << std::endl;
+        for (auto const & entry : transaction.get_gpg_signature_problems()) {
+            std::cerr << entry << std::endl;
+        }
+        for (auto & problem : transaction.get_transaction_problems()) {
+            std::cerr << "  - " << problem << std::endl;
+        }
+        throw libdnf5::cli::SilentCommandExitError(1);
+    }
+
+    for (auto const & entry : transaction.get_gpg_signature_problems()) {
+        std::cerr << entry << std::endl;
+    }
+
+    // If the upgrade succeeded, remove downloaded data
+    clean(ctx, get_datadir());
+
+    if (state.get_data().poweroff_after) {
```

Comment:
And when `poweroff_after` is false, the system is automatically rebooted?

Reply:
Ah, thanks, systemd _will_ do this for you but according to the docs, we are expected to trigger the reboot ourselves:


https://www.freedesktop.org/software/systemd/man/latest/systemd.offline-updates.html:
> The update scripts should exit only after the update is finished. It is expected that the service which performs the update will cause the machine to reboot after it is done. If the system-update.target is successfully reached, i.e. all update services have run, and the /system-update or /etc/system-update symlink still exists, it will be removed and the machine rebooted as a safety measure.